doc/developer/design/20210715_table_update.md

```diff
@@ -1,52 +1,158 @@
+state prior to merging this code.
+
+We want to prevent the case where one user opens a write transaction and forces
+all other write transactions for the first user to issue a `COMMIT`. The design
+above achieves this because `INSERT` only needs to serialize itself after the
+user has issued a `COMMIT` statement, and `ReadThenWrite` has no user
+interaction between its read and write. Additionally, `ReadThenWrite` must be
+excluded from explicit/multi-statement transactions.
+
+### Implementing global write lock
+
+In the initial design of this code, we relied on keeping track of the connection
+ID of the session that held the write lock as an `Option<u32>`. If the write
+lock was...
+- `Some`, only the connection with that ID was allowed to enter the critical
+  sections; others had to enqueue themselves and wait to be dequeued by the
+  lock's owner
+- `None`, entering the critical section saved the connection's ID as the
+  connection holding the lock.
+
+When the connection holding the lock was leaving the critical section, it was
+responsible for dequeuing an operation that was awaiting the write lock.
+
+A successive commit to this code, instead refactors the global write lock to use
+a `Mutex` instead of an `Option<u32>` because a `Mutex` offers exactly the
+semantics we want to serialize writes:
+- It limits a critical path to one entrant at a time
+- Signals those waiting on the lock of its availability and lets one of them
+    proceed
+
+While the mutex I chose comes from `tokio::sync`, mutexes are well-known,
+well-tested, idiomatic tools from standard libraries of many programming
+languages, so using one has the advantage of giving us exactly what we want with
+very little work, and is generally less fallible than trying to hand-roll the
+same semantics using another implemenation.
+
+By using a mutex, if the session holding the write lock gets dropped, the lock
+is released. To reach parity with the `Option<u32>` approach, we would need to
+implement an "unlock" operation on `Drop`, at which point we're basically
+handrolling a mutex.
+
+#### Considerations
+
+**Implementation**
+
+I chose `tokio::sync::Mutex` because its
+[documentation](https://docs.rs/tokio/1.11.0/tokio/sync/struct.Mutex.html) states:
+
+> [`tokio::sync::Mutex`'s] lock guard is designed to be held across `.await`
+> points
+
+And we want to hold this lock across at least one `await` point during commits.
+There are more details in the documentation.
+
+While this has a potential performance implication, performance is expressly not
+a goal of this code.
+
+**Where to put the mutex handle**
+
+Because e.g. `ReadThenWrite` nodes are "long-lived" calls, the simplest place to
+share data (i.e. the mutex's handle) for the life of the function call is on the
+`Session`. However, because `Session` is even-longer-lived, this means that
+there is not a naturally occurring place for the mutex to release itself, so you
```

Comment:
Could this go on the `Transaction`? It gets dropped at exactly the time when we would want to requeue something else. Might be weird because that struct only exists on certain transaction states, but maybe worth a look?